### ISSUE TITLE
Inaccuracy with XGBoost scoring using RData model object and replicating with JSON output

### ISSUE THREAD

**Issue by b-stats** [dmlc/xgboost#5183]
I am seeing something interesting with regards to replicating the scoring process of an XGBoost GBM. I have used R’s XGBoost functionality to train and score a model and am trying to use the JSON dump of the model object to score the model outside of the XGBoost environment.

I wrote Python code to replicate the scoring process using the outputted JSON file and was able to achieve 100% match between XGBoost’s R model object and the XGBoost outputted JSON file. However, when I changed some of the hyperparameters my match rate dropped. The main parameter that I changed was the learning rate which combined with early stopping results in more trees for larger learning rates. Below are the match rates that I see between the scoring directly from XGBoost and with the JSON object scoring:

**ETA (learning rate)**:   &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; .04 &nbsp;&nbsp;&nbsp;&nbsp;- &nbsp;&nbsp;&nbsp;.05 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- &nbsp;&nbsp;&nbsp;.06 &nbsp;&nbsp;&nbsp;&nbsp;- &nbsp;&nbsp;&nbsp;.07 &nbsp;&nbsp;&nbsp;- &nbsp;&nbsp;&nbsp;&nbsp;.1 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- &nbsp;&nbsp;&nbsp;&nbsp;.2   

**Match % (JSON vs RData)**: 100%&nbsp; -&nbsp; 99.9% &nbsp;- &nbsp;99.3% &nbsp;- &nbsp;98.9% - &nbsp;96.7%&nbsp; - &nbsp;94.8%&nbsp;

I am hypothesizing that there is a rounding issue that becomes apparent as the number of trees increases. I looked through some of the XGBoost source code and found the following piece of code which may be the culprit (https://github.com/dmlc/xgboost/blob/master/src/tree/tree_model.cc#L29-L35).
 But I can’t definitively say that this is the reason for the mismatch.

Does anyone have any ideas as to the reason for the change in match rates? Is it due to rounding or is there another reason for the discrepancy?

Appreciate any help that can be provided! Please let me know if I need to give any additional details.

**Comment by RAMitchell** [comment]
Did you add the base_score parameter when predicting?

**Comment by b-stats** [comment]
@RAMitchell yes, I have done it with and without the base_score set (which I am setting to the default value of .5 explicitly)

**Comment by trivialfis** [comment]
Will look into it, right now the JSON is still experimental.  Numeric IO procedures needs some more commits to both speed and accuracy.  The rounding guarantee from std is quite weak.

**Comment by b-stats** [comment]
Thanks for the update @trivialfis ! That is good to know. Please let me know if there is any additional information that I can provide.

**Comment by hcho3** [comment]
Closing now, since we now have a precise method for converting floating-point numbers to strings.
